### PR TITLE
Change uint256 format to big endian

### PR DESCRIPTION
**Pull request by thebevrishot**
closes https://github.com/zcoinofficial/ztm/issues/172

### CODE REVIEW

**Comment by thebevrishot** [comment]
After personal discussion, We can merge it without data migration script if we can finish this before first release.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,6 +25,13 @@ jobs:
     - name: Build E2E Runner
       shell: cmd
       run: dotnet.exe build src\Ztm.EndToEndTests
+    - name: Install main database
```
Comment (ultimaweapon) [inline .github/workflows/ci.yml]: I think this name quite hard to understand due to there is possibility that PostgreSQL will host additional database. So change this to `Install PostgreSQL` looks better.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -23,6 +25,13 @@ jobs:
...
+      shell: cmd
+      run: choco.exe install postgresql --version=10.6.0.20190131 --params '/Password:postgres' --params-global
```
Comment (ultimaweapon) [inline .github/workflows/ci.yml]: We should disable installation progress due to it output too much.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -23,6 +25,13 @@ jobs:
...
+    - name: Migrate main database
```
Comment (ultimaweapon) [inline .github/workflows/ci.yml]: Same here. My idea is `Create Main Database` looks better due to we did not migrate anything.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -23,6 +25,13 @@ jobs:
...
+      shell: cmd
+      working-directory: ./src/Ztm.Data.Entity.Postgres
```
Comment (ultimaweapon) [inline .github/workflows/ci.yml]: Path separator in Windows should be `\`. Also I think we can omit `.` at the beginning.

File: src/Ztm.Data.Entity.Postgres.Tests/TheoryWhenEnvIsSetAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+using System;
+using Xunit;
+
+namespace Ztm.Data.Entity.Postgres.Tests
+{
+    public class TheoryWhenEnvIsSetAttribute : TheoryAttribute
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity.Postgres.Tests/TheoryWhenEnvIsSetAttribute.cs]: IMO class name look weird. My idea is rename it to `ConditionalTheoryAttribute` with `RequiredEnv` property to specify the property name.

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseTests.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                }
+
+                // Get uint256 back.
+                var block = await this.subject.Blocks.FirstOrDefaultAsync();
+                Assert.NotNull(block);
+                Assert.Equal(parsed, block.Hash);
+            }
+            finally
+            {
+                this.subject.Database.ExecuteSqlCommand("TRUNCATE TABLE \"Blocks\" CASCADE");
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseTests.cs]: We should create a database fixture and move this to it instead.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (thebevrishot) [reply .github/workflows/ci.yml]: I'm not sure this field comply the Windows directory format and can't see the spec. However my Visual Studio Code show error message when I use Windows's format. This is the error message `String does not match the pattern of "^(\.[^\\?%*:|"<>\n]+)$".`

File: src/Ztm.Data.Entity.Postgres.Tests/TheoryWhenEnvIsSetAttribute.cs (hunk 5 above)
Comment (thebevrishot) [reply src/Ztm.Data.Entity.Postgres.Tests/TheoryWhenEnvIsSetAttribute.cs]: Good idea

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseTests.cs (hunk 6 above)
Comment (thebevrishot) [reply src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseTests.cs]: Done

File: .github/workflows/ci.yml (hunk 4 above)
Comment (ultimaweapon) [reply .github/workflows/ci.yml]: Have you tried this? `src\Ztm.Data.Entity.Postgres`

Comment (thebevrishot) [reply .github/workflows/ci.yml]: It show the same error

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        public void CleanUp()
+        {
+            // Truncate all tables in schema public except __EFMigrationsHistory
+            var tables = this.ExecuteSql("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public' AND table_name <> '__EFMigrationsHistory';")
+                             .Select(r => r[0] as string);
+
+            using (var db = this.CreateMainDatabase())
+            {
+                var query = "TRUNCATE TABLE " + string.Join(", ", tables.Select(t => '"' + t + '"'));
+                db.Database.ExecuteSqlCommand(query); // lgtm [cs/second-order-sql-injection]
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs]: I think we should use `DbConnection` directly for executing raw SQL.

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+        public void Dispose()
+        {
+            this.connection.Dispose();
+        }
+
+        public IEnumerable<IDataRecord> ExecuteSql(string rawSql)
+        {
+            using (var db = this.CreateMainDatabase())
+            using (var command = db.Database.GetDbConnection().CreateCommand())
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs]: Same here. We can create `DbCommand` from `DbConnection` directly.

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                db.Database.OpenConnection();
+                using (var result = command.ExecuteReader())
+                {
+                    while (result.Read())
+                    {
+                        yield return result;
+                    }
+                }
+
+                db.Database.CloseConnection();
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs]: Seems like this one is going to close underlying connection, which mean is will cause subsequence operation to fail due to we have only one connection.

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFactoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            act = () => new MainDatabaseFactory(null);
+            act.Should().Throw<ArgumentNullException>()
+               .And.ParamName.Should().Be("config");
+        }
+
+        [Fact]
+        public void CreateDbContext_ShouldSuccess()
+        {
+            // Act.
+            var context = this.subject.CreateDbContext();
```
Comment (ultimaweapon) [inline src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFactoryTests.cs]: We forgot to dispose this.

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs (hunk 9 above)
Comment (thebevrishot) [reply src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs]: That's right.

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs (hunk 7 above)
Comment (thebevrishot) [reply src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs]: Good idea

File: src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+        public MainDatabaseFixture()
+        {
+            var connectionString = Environment.GetEnvironmentVariable("ZTM_MAIN_DATABASE");
+            if (connectionString == null)
+            {
+                throw new Exception("No ZTM_MAIN_DATABASE environment variable is set.");
+            }
+
+            this.connection  = new NpgsqlConnection(connectionString);
```
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/Ztm.Data.Entity.Postgres.Tests/MainDatabaseFixture.cs]: fixed

**Pull request merged**